RenderObject::isOutOfFlowPositioned can be read before it gets set.

mattwoodrow opened the pull request:
#### 055d5a97c962454efc1124c46f7cac5b35812164
<pre>
RenderObject::isOutOfFlowPositioned can be read before it gets set.
<a href="https://bugs.webkit.org/show_bug.cgi?id=287778">https://bugs.webkit.org/show_bug.cgi?id=287778</a>
&lt;<a href="https://rdar.apple.com/144956226">rdar://144956226</a>&gt;

Reviewed by Alan Baradlay.

RenderBoxModelObject::updateFromStyle updates the cached position flags using
the new RenderStyle object.

RenderElement::styleDidChange (which happens earlier) can schedule layout, which
depends on the state of these flags.

Call updateFromStyle to update any cached state before continuing to handle the
post-style-change code.

As a result of this change, `RenderElement::styleDidChange` ancestor marking now
correctly sets the &apos;positioned child needs layout&apos; bit on the container for
changed positioned children, rather than always setting &apos;normal child needs
layout&apos;. This results in simplified layout happening more often.

This exposes a bug where changes to the display property may require static
positions to be recomputed, which doesn&apos;t get done with simplified layout. Adds
a check for this to ensure we don&apos;t use simplified layout for this subset.

This also exposes a second bug with intrusive scrollbars, where overflow changes
on positioned children may results in the addition/removal of scrollbars and
require a full layout to adjust to the new size. Adds checks for this at the two
entry points for simplified layout and switches back to regular layout requests.

Enables two existing tests (which were incorrectly disabled) that caught these
two bugs.

Since we now sometimes refuse simplified layout upfront, rename
setNeedsSimplifiedNormalFlowLayout to setNeedsLayoutForOverflowChange, and
StyleDifference::SimplifiedLayout to Overflow to better represent the request
being made (and have the details of what layout type happens be internal).

Deduplicate some StyleDifference -&gt; layout request code into
setNeedsLayoutForStyleDifference.

Remove markFragmentsForOverflowLayoutIfNeeded since it was unused.

* LayoutTests/TestExpectations:
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::overflowChangesMayAffectLayout const):
* Source/WebCore/rendering/RenderBox.h:
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::adjustStyleDifference const):
(WebCore::RenderElement::repaintBeforeStyleChange):
(WebCore::RenderElement::setStyle):
(WebCore::RenderElement::styleDidChange):
(WebCore::RenderElement::setNeedsLayoutForStyleDifference):
(WebCore::RenderElement::setNeedsLayoutForOverflowChange):
(WebCore::RenderElement::setNeedsSimplifiedNormalFlowLayout): Deleted.
* Source/WebCore/rendering/RenderElement.h:
(WebCore::RenderElement::overflowChangesMayAffectLayout const):
* Source/WebCore/rendering/RenderFragmentedFlow.cpp:
(WebCore::RenderFragmentedFlow::markFragmentsForOverflowLayoutIfNeeded): Deleted.
* Source/WebCore/rendering/RenderFragmentedFlow.h:
* Source/WebCore/rendering/RenderLayerModelObject.cpp:
(WebCore::RenderLayerModelObject::styleDidChange):
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::markContainingBlocksForLayout):
* Source/WebCore/rendering/style/RenderStyleConstants.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/rendering/style/RenderStyleConstants.h:
* Source/WebCore/rendering/updating/RenderTreeUpdater.cpp:
(WebCore::repaintAndMarkContainingBlockDirtyBeforeTearDown):

Canonical link: <a href="https://commits.webkit.org/290831@main">https://commits.webkit.org/290831@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f9a31ed221e8dc8da7ff9ce4eff917f73cfe1730

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/91101 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/10644 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/136 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/96112 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/41868 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/93151 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/11050 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/18962 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/70019 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27540 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/94102 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/8425 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/82550 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/50345 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/8196 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/116 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/40998 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/78518 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/133 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/98088 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/18305 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/13426 "Found 1 new test failure: accessibility/mac/iframe-pdf.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/79029 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/18564 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/78375 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/78231 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19363 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/22730 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/87 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/11514 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/18306 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/23638 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/18032 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/21493 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/19811 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/rendering/RenderLayerModelObject.cpp:
```diff
@@ -133,8 +133,8 @@ void RenderLayerModelObject::styleWillChange(StyleDifference diff, const RenderS
-    RenderElement::styleDidChange(diff, oldStyle);
     updateFromStyle();
+    RenderElement::styleDidChange(diff, oldStyle);
 
```
I am surprised no code relies on this odd state where style is already set on the renderer but caches haven't been populated yet. I guess if something starts failing that should probably go to styleWillChange anyway.
Also why isn't it in RenderElement! RenderLayerModelObject sounds like a random place for a function like "update from style" -though I assume it is driven by actual use cases.

alanbaradlay commented on Source/WebCore/rendering/RenderElement.cpp:
```diff
@@ -1068,6 +1068,11 @@ void RenderElement::styleDidChange(StyleDifference diff, const RenderStyle* oldS
         setNeedsPositionedMovementLayout(oldStyle);
 
+    if (isOutOfFlowPositioned() && oldStyle && oldStyle->isOriginalDisplayBlockType() != style().isOriginalDisplayBlockType()) {
+        if (CheckedPtr ancestor = container())
```
I think we want containing block here.

mattwoodrow replied:
That's a good point.

Is `RenderObject::markContainingBlocksForLayout` incorrect? It seems like it walks up the container chain, and would mark more intermediate ancestors than is necessary (especially for position:fixed).

alanbaradlay replied:
In some cases maybe. In general though you want continuity (no gaps in the ancestor chain) so that's why we have container() here (as opposed to containing block)...but I just realized that this is about static positioning which is computed as if the out-of-flow box was in-flow so both container and containing block would return incorrect ancestor here.

alanbaradlay replied:
what we need is the in-flow containing block, which is usually the parent, except when the parent is an inline box e.g. `<div><span><div out-of-flow>`

mattwoodrow committed:
RenderObject::isOutOfFlowPositioned can be read before it gets set.
https://bugs.webkit.org/show_bug.cgi?id=287778
<rdar://144956226>

Reviewed by Alan Baradlay.

RenderBoxModelObject::updateFromStyle updates the cached position flags using
the new RenderStyle object.

RenderElement::styleDidChange (which happens earlier) can schedule layout, which
depends on the state of these flags.

Call updateFromStyle to update any cached state before continuing to handle the
post-style-change code.

As a result of this change, `RenderElement::styleDidChange` ancestor marking now
correctly sets the 'positioned child needs layout' bit on the container for
changed positioned children, rather than always setting 'normal child needs
layout'. This results in simplified layout happening more often.

This exposes a bug where changes to the display property may require static
positions to be recomputed, which doesn't get done with simplified layout. Adds
a check for this to ensure we don't use simplified layout for this subset.

This also exposes a second bug with intrusive scrollbars, where overflow changes
on positioned children may results in the addition/removal of scrollbars and
require a full layout to adjust to the new size. Adds checks for this at the two
entry points for simplified layout and switches back to regular layout requests.

Enables two existing tests (which were incorrectly disabled) that caught these
two bugs.

Since we now sometimes refuse simplified layout upfront, rename
setNeedsSimplifiedNormalFlowLayout to setNeedsLayoutForOverflowChange, and
StyleDifference::SimplifiedLayout to Overflow to better represent the request
being made (and have the details of what layout type happens be internal).

Deduplicate some StyleDifference -> layout request code into
setNeedsLayoutForStyleDifference.

Remove markFragmentsForOverflowLayoutIfNeeded since it was unused.

* LayoutTests/TestExpectations:
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::overflowChangesMayAffectLayout const):
* Source/WebCore/rendering/RenderBox.h:
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::adjustStyleDifference const):
(WebCore::RenderElement::repaintBeforeStyleChange):
(WebCore::RenderElement::setStyle):
(WebCore::RenderElement::styleDidChange):
(WebCore::RenderElement::setNeedsLayoutForStyleDifference):
(WebCore::RenderElement::setNeedsLayoutForOverflowChange):
(WebCore::RenderElement::setNeedsSimplifiedNormalFlowLayout): Deleted.
* Source/WebCore/rendering/RenderElement.h:
(WebCore::RenderElement::overflowChangesMayAffectLayout const):
* Source/WebCore/rendering/RenderFragmentedFlow.cpp:
(WebCore::RenderFragmentedFlow::markFragmentsForOverflowLayoutIfNeeded): Deleted.
* Source/WebCore/rendering/RenderFragmentedFlow.h:
* Source/WebCore/rendering/RenderLayerModelObject.cpp:
(WebCore::RenderLayerModelObject::styleDidChange):
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::markContainingBlocksForLayout):
* Source/WebCore/rendering/style/RenderStyleConstants.cpp:
(WebCore::operator<<):
* Source/WebCore/rendering/style/RenderStyleConstants.h:
* Source/WebCore/rendering/updating/RenderTreeUpdater.cpp:
(WebCore::repaintAndMarkContainingBlockDirtyBeforeTearDown):

Canonical link: https://commits.webkit.org/290831@main
